vis_methods.py

```diff
@@ -0,0 +1,271 @@
+#
+# Tehuti is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Tehuti.  If not, see <http://www.gnu.org/licenses/>.
+import warnings
+
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy.stats import gaussian_kde
+
+from tehuti import shorten_sha
+
+
+Y_AXIS_LABELS = {'timeit': 'Time (s)',
+                 'linecount': 'Number of lines',
+                 'pylint': 'PyLint score',
+                 'memoryuse': 'Memory (MB)',
+                 'accuracy': 'Accuracy (%)'}
+
+
+class Visualiser(object):
+    """The base class for visualiser states."""
+
+    def select_data(self, commit, metrics):
+        if commit is None:
+            commits = self.vis.results.keys()
+        elif isinstance(commit, basestring):
+            commits = list(commit)
+        if metrics is None:
+            # Only choose metrics that show up in all runs.
+            keys = None
+            for metrics in self.vis.results.values():
+                if keys is None:
+                    keys = set(metrics.keys()) - {'name'}
+                else:
+                    metrics_keys = metrics.keys()
+                    common = set(metrics_keys) & keys
+                    keys = common
+            metrics = list(keys)
+        return commits, metrics
+
+    def plot(self, alternate_plot):
+        raise NotImplemented
+
+
+class VaryRepoCommit(Visualiser):
+    """
+    A visualiser state that plots benchmark results against repository commits
+    for all supplied metrics.
+
+    Metric results may either be plotted each on individual axes, or all on the
+    same axes.
+
+    """
+    def __init__(self, vis):
+        self.vis = vis
+
+    def select_data(self, commits, metrics):
+        """
+        The select data method for this state.
+
+        Select and re-format data from the supplied metrics results file.
+
+        Data is selected based on all specified `commits` and `metrics`. Data
+        is formatted into the standard format used by plot states, with
+        specific formatting that collapses any lists of values into the single
+        minimum value in that list.
+        If no commits or metrics are specified then all commits or metrics
+        in the supplied metrics results file are selected.
+
+        Args:
+
+        * commit:
+            One or more valid repository commits that have been benchmarked.
+        * metrics:
+            One or more metrics that have results in the metrics results file.
+
+        Returns:
+            The selected data formatted for plotting.
+
+        """
+        commits, metrics = super(VaryRepoCommit, self).select_data(commits,
+                                                                   metrics)
```

Comment:
@dkillick Oh, be careful here with this pattern ... this `select_data` is specializing with the same signature as its parent, but they return different things i.e. the parent returns a tuple of `commits, metrics` whereas the sub-class returns just `data` ... you may want to rename the common parent method? It's a subtle difference ...